src/v/cluster/cloud_metadata/cluster_recovery_backend.cc

```diff
@@ -0,0 +1,427 @@
+    auto fib = retry_chain_node{_as, 30s, 1s};
+    if (!_recovery_table.local().is_recovery_active()) {
+        co_return std::nullopt;
+    }
+    auto recovery_state
+      = _recovery_table.local().current_recovery().value().get();
+    const auto& cluster_manifest = recovery_state.manifest;
+
+    // Download the controller snapshot.
+    const auto& controller_path_str = cluster_manifest.controller_snapshot_path;
+    if (cluster_manifest.controller_snapshot_path.empty()) {
+        co_return std::nullopt;
+    }
+    vlog(
+      clusterlog.info,
+      "Using controller snapshot at remote path {} in bucket {}",
+      controller_path_str,
+      bucket);
+    auto remote_controller_snapshot = cloud_storage::remote_file(
+      _remote,
+      _cache,
+      bucket,
+      cloud_storage::remote_segment_path{controller_path_str},
+      fib,
+      "controller_snapshot");
+
+    try {
+        auto f = co_await remote_controller_snapshot.hydrate_readable_file();
+        ss::file_input_stream_options options;
+        auto input = ss::make_file_input_stream(f, options);
+        storage::snapshot_reader reader(
+          std::move(f),
+          std::move(input),
+          remote_controller_snapshot.local_path());
+
+        // Parse the snapshot, and make sure to close the snapshot reader
+        // before destructing it, even on failure.
+        std::exception_ptr eptr;
+        cluster::controller_snapshot snapshot;
+        try {
+            auto snap_metadata_buf = co_await reader.read_metadata();
+            auto snap_metadata_parser = iobuf_parser(
+              std::move(snap_metadata_buf));
+            auto snap_metadata = reflection::adl<raft::snapshot_metadata>{}
+                                   .from(snap_metadata_parser);
+            const size_t snap_size = co_await reader.get_snapshot_size();
+            auto snap_buf_parser = iobuf_parser{
+              co_await read_iobuf_exactly(reader.input(), snap_size)};
+            snapshot = serde::read<cluster::controller_snapshot>(
+              snap_buf_parser);
+        } catch (...) {
+            eptr = std::current_exception();
+        }
+        co_await reader.close();
+        if (eptr) {
+            std::rethrow_exception(eptr);
+        }
+        co_return snapshot;
+    } catch (...) {
+        vlog(
+          clusterlog.warn,
+          "Error processing controller snapshot: {}",
+          std::current_exception());
+        co_return std::nullopt;
+    }
+}
+
+ss::future<> cluster_recovery_backend::recover_until_abort() {
+    co_await _features.await_feature(
+      features::feature::cloud_metadata_cluster_recovery, _as);
+    while (!_as.abort_requested()) {
+        auto& recovery_table = _recovery_table.local();
+        co_await recovery_table.wait_for_active_recovery();
+        if (recovery_table.is_recovery_active()) {
+            if (!_raft0->is_leader()) {
+                bool shutdown = false;
```

Comment:
nit: i think this is not required as `_as.request_abort` and `_leader_cond.broken()` are called in one synchronous block